vllm/entrypoints/openai/protocol.py

```diff
@@ -454,6 +454,12 @@ def check_tool_usage(cls, data):
         if "tool_choice" not in data and data.get("tools"):
             data["tool_choice"] = "auto"
 
+        # if "tool_choice" is "none" -- ignore tools if present
+        if "tool_choice" in data and data["tool_choice"] == "none":
+            # ensure that no tools are present
+            data.pop("tools", None)
+            return data
+
```

Comment:
This is a good solution. 

Worth noting, the model could still hallucinate a tool call though (smaller models do this on occasion; I have seem llama and some fine-tunes of it do this), and the tool parser would still parse it and return an assistant-role tool call chat completion.

It should be possible to disable tool parsers for the duration of a request or to bypass them so that even if the model does "try" to call a tool by hallucinating one, it will be returned as an assistant-role text chat completion rather than a tool call chat completion.

This may be overkill. What do you think @gcalmettes @DarkLight1337 @mgoin ?

Reply:
Otherwise LGTM :)

Reply:
> It should be possible to disable tool parsers for the duration of a request or to bypass them so that even if the model does "try" to call a tool by hallucinating one, it will be returned as an assistant-role text chat completion rather than a tool call chat completion.

How would we detect that?

Reply:
I don't think we need to detect anything, we just disable tool parsing & extraction if the request specifies `tool_choice="none"` in addition to not passing tools to the chat template, so that any tokens generated whether they are text or a hallucinated tool call are processed as normal text due to the tool parser being disabled.

Reply:
From my understanding, we should be able to achieve this by updating this code here: https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/serving_engine.py#L473

So that the tool parser isn't even called if `tool_dicts is None`.

Reply:
@DarkLight1337 @K-Mistele I agree, preventing any tool parsing if `tool_choice` is set to `none` would be nice to prevent hallucinated `tool_calls` to be parsed.
I have added a commit for deactivating the `adjust_request` flow when `tool_choice` is `none` as you suggested  @DarkLight1337

Reply:
and to bounce back on what you were mentioning @K-Mistele , we are definitely observing the same behavior with Llama 8B (quantized or not), whereas we haven’t seen it yet with Llama 70B or any small Mistral models 

> Worth noting, the model could still hallucinate a tool call though (smaller models do this on occasion; I have seem llama and some fine-tunes of it do this), and the tool parser would still parse it and return an assistant-role tool call chat completion.

Reply:
We also need to deactivate calls to `extract_tool_calls` and `extract_tool_calls_streaming`.

Reply:
@DarkLight1337 from my understanding, the cases of skipping `extract_tool_calls` and `extract_tool_calls_streaming` are already handled:
- for non streaming case, if `tool_choice` is `none` the generated LLM output is [already directly set without tool extraction](https://github.com/vllm-project/vllm/blob/8a4358ecb5ba457fad2be0ed930132489eddddf5/vllm/entrypoints/openai/serving_chat.py#L636) (so the tool extraction process, [here](https://github.com/vllm-project/vllm/blob/8a4358ecb5ba457fad2be0ed930132489eddddf5/vllm/entrypoints/openai/serving_chat.py#L652), is not run in this case)
- similarly in streaming, [this check](https://github.com/vllm-project/vllm/blob/8a4358ecb5ba457fad2be0ed930132489eddddf5/vllm/entrypoints/openai/serving_chat.py#L412) does not match `tool_calls="none"`

Reply:
Ah, I see. Should be good then.